File: languages/go/03-advanced/inheritance-and-polymorphism/example/polymorphic-menu.go
// This extra example extends inheritance-and-polymorphism with a polymorphic menu.
// Example purpose: collect different shapes behind one shared interface.

package main

import "fmt"

type shape interface {
	area() float64
	name() string
}

type circle struct {
	radius float64
}

func (current circle) area() float64 {
	return 3.1415926535 * current.radius * current.radius
}

func (current circle) name() string {
	return "Circle"
}

type rectangle struct {
	width  float64
	height float64
}

func (current rectangle) area() float64 {
	return current.width * current.height
}

func (current rectangle) name() string {
	return "Rectangle"
}

func main() {
	shapes := make([]shape, 0)

	for {
		fmt.Println("\nMenu")
		fmt.Println("1) Add circle")
		fmt.Println("2) Add rectangle")
		fmt.Println("3) List areas")
		fmt.Println("4) Exit")
		fmt.Print("Choose: ")

		var option int
		if _, err := fmt.Scan(&option); err != nil {
			fmt.Println("Invalid option.")
			return
		}

		switch option {
		case 1:
			var radius float64
			fmt.Print("Radius: ")
			if _, err := fmt.Scan(&radius); err != nil || radius <= 0.0 {
				fmt.Println("Radius must be positive.")
				continue
			}
			shapes = append(shapes, circle{radius: radius})
		case 2:
			var width float64
			var height float64
			fmt.Print("Width: ")
			if _, err := fmt.Scan(&width); err != nil {
				fmt.Println("Width and height must be positive.")
				continue
			}
			fmt.Print("Height: ")
			if _, err := fmt.Scan(&height); err != nil || width <= 0.0 || height <= 0.0 {
				fmt.Println("Width and height must be positive.")
				continue
			}
			shapes = append(shapes, rectangle{width: width, height: height})
		case 3:
			if len(shapes) == 0 {
				fmt.Println("No shapes added yet.")
				continue
			}

			totalArea := 0.0
			for _, current := range shapes {
				shapeArea := current.area()
				totalArea += shapeArea
				fmt.Printf("- %s area: %v\n", current.name(), shapeArea)
			}
			fmt.Printf("Total area: %v\n", totalArea)
		case 4:
			fmt.Println("Goodbye.")
			return
		default:
			fmt.Println("Invalid option.")
		}
	}
}

File: languages/cpp/03-advanced/inheritance-and-polymorphism/example/polymorphic-menu.cpp
// This helper example focuses on isolating one polymorphic dispatch path so the common interface
// stays clear.

#include <iostream>
#include <memory>
#include <vector>

using namespace std;

class Shape {
  public:
    virtual ~Shape() = default;
    virtual double area() const = 0;
    virtual const char* name() const = 0;
};

class Circle : public Shape {
  public:
    explicit Circle(double radiusValue) : radius(radiusValue) {}

    double area() const override { return 3.1415926535 * radius * radius; }

    const char* name() const override { return "Circle"; }

  private:
    double radius;
};

class Rectangle : public Shape {
  public:
    Rectangle(double widthValue, double heightValue) : width(widthValue), height(heightValue) {}

    double area() const override { return width * height; }

    const char* name() const override { return "Rectangle"; }

  private:
    double width;
    double height;
};

int main() {
    vector<unique_ptr<Shape>> shapes;

    while (true) {
        int option = 0;
        cout << "\nMenu\n";
        cout << "1) Add circle\n";
        cout << "2) Add rectangle\n";
        cout << "3) List areas\n";
        cout << "4) Exit\n";
        cout << "Choose: ";
        cin >> option;

        if (option == 1) {
            double radius = 0.0;
            cout << "Radius: ";
            cin >> radius;
            if (radius > 0.0) {
                shapes.push_back(make_unique<Circle>(radius));
            } else {
                cout << "Radius must be positive.\n";
            }
        } else if (option == 2) {
            double width = 0.0;
            double height = 0.0;
            cout << "Width: ";
            cin >> width;
            cout << "Height: ";
            cin >> height;
            if (width > 0.0 && height > 0.0) {
                shapes.push_back(make_unique<Rectangle>(width, height));
            } else {
                cout << "Width and height must be positive.\n";
            }
        } else if (option == 3) {
            if (shapes.empty()) {
                cout << "No shapes added yet.\n";
                continue;
            }

            double totalArea = 0.0;
            for (const unique_ptr<Shape>& shape : shapes) {
                const double shapeArea = shape->area();
                totalArea += shapeArea;
                cout << "- " << shape->name() << " area: " << shapeArea << '\n';
            }
            cout << "Total area: " << totalArea << '\n';
        } else if (option == 4) {
            cout << "Goodbye.\n";
            break;
        } else {
            cout << "Invalid option.\n";
        }
    }

    return 0;
}
